### PR TITLE
Sylvia whittle/gwy file format

### LINKED ISSUE #414: Support GWY formats
Further to #137 it would be useful to be able to load and process [Gwyddion](http://gwyddion.net/) native file format in TopoStats.

The file format is described [here](http://gwyddion.net/documentation/user-guide-en/gwyfile-format.html) and it is possible the Python 3.0 library [gwyfile](https://github.com/tuxu/gwyfile) will ease the effort required to implement this.

### PR DESCRIPTION
**Pull request by SylviaWhittle**
Closes #414 

This PR adds support for `.gwy` files.
- Adds several binary file reading functions.
- Adds a development-only function (`_gwy_print_dict` and `_gwy_print_dict_wrapper`) that helps in exploring the `.gwy` file format. I would recommend keeping this in, as it can be very difficult to find what keys are available in the nested object-component dictionary.
- Adds the `load_gwy()` function that loads the `.gwy` file and returns the image and `pixel_to_nm_scaling_factor`.

### CODE REVIEW

**Review by ns-rse — COMMENTED** [review]
Excellent work digging into binary files. :+1: 

First pass review, we've already chatted about the need to test `.gwy` files derived from sources other than `.spm` and I know you found an issue with `.ibw` files and are working through that and adding tests using the files under `tests/resources/`.

File: topostats/validation.py
Comment (SylviaWhittle) [reply topostats/validation.py:48]: The `.gwy` file format does not seem to have channels. It appears to be structured so that the first object within the initial object contains a component with the heightmap data. There isn't really a channel name associated with it aside from the component name of 'data'.

File: topostats/io.py
Comment (SylviaWhittle) [reply topostats/io.py:177]: Unit tests for the methods in `io.py` have already been written, see lines 58 - 99 and 224 - 259 of `test_io.py`

File: topostats/io.py
Comment (SylviaWhittle) [reply topostats/io.py:199]: I was unsure about this myself too. My reasoning for putting it in the base `io.py` is that the functions aren't inherent to loading scans, and are more of general I/O than loading scans. Why would we want to force a LoadScans instance to be created when the functions could just be standalone?

**However** I will gladly defer to your judgement on this if you think they should go in LoadScans.

File: topostats/io.py
Comment (SylviaWhittle) [reply topostats/io.py:600]: I loved this suggestion, the `json.dumps` method is so elegant, however when I went to try it...

```
000nm.0_00060.gwy
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/TopoStats/topostats/run_topostats.py", line 189, in main
    all_scan_data.get_data()
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 705, in get_data
    self.image, self.pixel_to_nm_scaling = self.load_gwy()
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 649, in load_gwy
    LoadScans._gwy_print_dict(gwy_file_dict=image_data_dict, indent=2, sort_keys=False)
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 625, in _gwy_print_dict
    LOGGER.info(json.dumps(gwy_file_dict, indent=indent, sort_keys=sort_keys))
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ndarray is not JSON serializable
```

And from what I have read, each numpy array would need to be turned into a list manually, so it would require a script like I have already written, recursively turning arrays to lists...

Comment (ns-rse) [reply topostats/io.py:600]: That is mildly annoying. I'll go through this PR again tomorrow.

File: topostats/io.py
Comment (ns-rse) [reply topostats/io.py:199]: Given this some thought and they probably shouldn't be methods of LoadScans, but then they don't warrant a class of their own either.

One option to keeping the code organised though would be to group them into a sub-module such as `topostats/io/binary.py` (`binary` because I can't think of a better name!). Would require adjusting the existing imports, moving tests and importing them to `topostats/io.py` though.

If you think this is worthwhile I'd say make it a separate issue and merge this then rejig things around, otherwise this PR gets dragged out.

File: topostats/io.py
Comment (ns-rse) [reply topostats/io.py:600]: There is a neat solution to this which involves converting arrays to lists on the fly (see [SO solution](https://stackoverflow.com/a/47626762/1444043)).

Not sure that is actually any more succinct, but defining a dedicated class as in the example would make the code more generalisable and it could then be used elsewhere if required.

As above perhaps a separate issue should be created otherwise this PR will start growing legs.

Comment (SylviaWhittle) [reply topostats/io.py:620]: Made issue #483 👍

**Review by ns-rse — APPROVED** [review]
Looks good to merge once the tests have finished. :+1:

**Pull request merged**
